File: backends/ts-hono/src/usecase/staff/interactor.ts
/**
 * スタッフユースケース Interactor モジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
import { badRequest, notFound } from "../../lib/errors.js";
import type { StaffRepository } from "../../domain/staff/repository.js";
import type { StaffListItem } from "../../domain/staff/valueObjects.js";
import { mapper } from "../../support/mapper.js";
import { StaffSymbol, StaffListItemSymbol } from "../../support/mappers/index.js";

/** スタッフのユースケース実装。 */
export class StaffInteractor {
  constructor(private readonly repo: StaffRepository) {}

  /**
   * 検索条件に合致するスタッフ一覧の VO と総件数を返します。
   * @param keyword - キーワード検索
   * @param roles - ロールフィルター
   * @param statuses - 状態フィルター（1=有効, 0=無効）
   * @param offset - オフセット
   * @param limit - 取得件数
   * @param sort - ソート対象
   * @param sortType - ソート順
   * @returns StaffListItem の配列と総件数のタプル
   */
  async findByCondition(keyword?: string, roles?: number[], statuses?: number[], offset = 0, limit = 10, sort?: string, sortType?: string): Promise<[StaffListItem[], number]> {
    const count = await this.repo.countAll(keyword, roles, statuses);
    const staffs = await this.repo.findAll(keyword, roles, statuses, { offset, limit, sort, sortType });
    return [mapper.mapArray(staffs, StaffSymbol, StaffListItemSymbol), count];
  }

  /**
   * スタッフの権限を更新します。
   * @param staffId - スタッフID
   * @param role - 新しい権限値
   * @param version - 楽観排他ロック用バージョン
   * @param executorId - 操作者スタッフID
   * @throws AppError 自分自身のロール更新、またはスタッフが存在しない場合
   */
  async updateRole(staffId: number, role: number, version: number, executorId: number): Promise<void> {
    if (staffId === executorId) throw badRequest("cannot_update_own_role");
    // 無効化（論理削除）はログイン可否にのみ影響するため、権限更新は無効スタッフも対象に含める。
    const staff = await this.repo.findByIdUnscoped(staffId);
    if (!staff) throw notFound("staff_not_found");
    await this.repo.updateRole(staffId, role, version);
  }

  /**
   * スタッフの論理削除を復元します。
   * @param staffId - スタッフID
   * @throws AppError スタッフが存在しない場合
   */
  async restore(staffId: number): Promise<void> {
    const staff = await this.repo.findByIdUnscoped(staffId);
    if (!staff) throw notFound("staff_not_found");
    await this.repo.restore(staffId, staff.version);
  }

  /**
   * スタッフを論理削除します。
   * @param staffId - スタッフID
   * @param version - 楽観排他ロック用バージョン
   * @param executorId - 操作者スタッフID
   * @throws AppError 自分自身の削除、またはスタッフが存在しない場合
   */
  async destroy(staffId: number, version: number, executorId: number): Promise<void> {
    if (staffId === executorId) throw badRequest("cannot_delete_self");
    const staff = await this.repo.findById(staffId);
    if (!staff) throw notFound("staff_not_found");
    await this.repo.softDelete(staffId, version);
  }
}

File: backends/python-fastapi/app/usecase/staff/interactor.py
"""
スタッフユースケース Interactor モジュール。

Author: Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
"""
from typing import Optional

from app.domain.staff.entity import Staff
from app.domain.staff.condition import StaffCondition
from app.domain.staff.repository import StaffRepository
from app.domain.staff.value_objects import StaffListItem
from app.exceptions import not_found, bad_request
from app.usecase.staff.dto import StaffUpdateRoleDto, StaffDestroyDto


def _to_list_item(staff: Staff) -> StaffListItem:
    """スタッフエンティティを一覧用 Vo に変換します。

    Args:
        staff: スタッフエンティティ

    Returns:
        StaffListItem インスタンス
    """
    status = 0 if staff.deleted_at is not None else 1
    return StaffListItem(
        id=staff.id,
        name=staff.name,
        email=staff.email,
        role=staff.role,
        status=status,
        created_at=staff.created_at,
        updated_at=staff.updated_at,
    )


class StaffInteractor:
    """スタッフのユースケース実装。

    Attributes:
        repository: スタッフリポジトリ
    """

    def __init__(self, repo: StaffRepository):
        """初期化します。

        Args:
            repo: スタッフリポジトリ
        """
        self.repository = repo

    def find_by_condition(
        self,
        keyword: Optional[str] = None,
        roles: Optional[list[int]] = None,
        offset: int = 0,
        limit: int = 10,
        sort: Optional[str] = None,
        sort_type: Optional[str] = None,
    ) -> tuple[list[StaffListItem], int]:
        """検索条件に合致するスタッフ一覧の Vo と総件数を返します。

        Args:
            keyword: キーワード検索文字列
            roles: ロールフィルター
            offset: オフセット
            limit: 取得件数
            sort: ソート対象
            sort_type: ソート順

        Returns:
            StaffListItem のリストと総件数のタプル
        """
        cond = StaffCondition(keyword=keyword, roles=roles or [], offset=offset, limit=limit, sort=sort, sort_type=sort_type)
        count = self.repository.count_staffs(cond)
        staffs = self.repository.find_all_staffs(cond)
        return [_to_list_item(s) for s in staffs], count

    def update_role(self, dto: StaffUpdateRoleDto) -> None:
        """スタッフの権限を更新します。

        Args:
            dto: ロール更新 Dto

        Raises:
            AppException: 自分自身のロール更新、またはスタッフが存在しない場合
        """
        if dto.staff_id == dto.executor_id:
            raise bad_request("cannot_update_own_role")
        staff = self.repository.find_staff_by_id(dto.staff_id)
        if staff is None:
            raise not_found("staff_not_found")
        self.repository.update_staff_role(staff, dto.role)

    def restore(self, staff_id: int) -> None:
        """スタッフの論理削除を復元します。

        Args:
            staff_id: スタッフID

        Raises:
            AppException: スタッフが存在しない場合
        """
        staff = self.repository.find_staff_by_id_include_deleted(staff_id)
        if staff is None:
            raise not_found("staff_not_found")
        self.repository.restore_staff(staff)

    def destroy(self, dto: StaffDestroyDto) -> None:
        """スタッフを論理削除します。

        Args:
            dto: 論理削除 Dto

        Raises:
            AppException: 自分自身の削除、またはスタッフが存在しない場合
        """
        if dto.staff_id == dto.executor_id:
            raise bad_request("cannot_delete_self")
        staff = self.repository.find_staff_by_id(dto.staff_id)
        if staff is None:
            raise not_found("staff_not_found")
        self.repository.soft_delete_staff(staff)
